src/k2/persistence/logStream/LogStream.cpp

```diff
@@ -0,0 +1,547 @@
+    K2LOG_D(log::lgbase, "Pre Allocated Plogs for LogStreaBase");
+    if (_create){
+        throw k2::dto::LogStreamBaseExistError("Log stream already created");
+    }
+    _switched = false;
+
+    // create metadata plogs in advance
+    std::vector<seastar::future<std::tuple<Status, String> > > createFutures;
+    for (uint32_t i = 0; i < PLOG_POOL_SIZE + 1; ++i){
+        createFutures.push_back(_client.create());
+    }
+    return seastar::when_all_succeed(createFutures.begin(), createFutures.end())
+    .then([this] (auto&& responses){
+        for (auto& response: responses){
+            auto& [status, plogId] = response;
+            if (!status.is2xxOK()){
+                throw k2::dto::PlogCreateError("unable to create plog for Logstream Base");
+            }
+            _preallocatedPlogPool.push_back(std::move(plogId));
+        }
+        return seastar::make_ready_future<>();
+    });
+}
+
+
+seastar::future<>
+LogStreamBase::_activeAndPersistTheFirstPlog(){
+    String plogId = _preallocatedPlogPool.back();
+    _preallocatedPlogPool.pop_back();
+    PlogInfo info{.currentOffset=0, .sealed=false, .next_plogId=""};
+    _first_plogId = plogId;
+    _current_plogId = plogId;
+    _usedPlogInfo[std::move(plogId)] = std::move(info);
+    
+    // persist this used plog info
+    return _addNewPlog(0, _current_plogId)
+    .then([this] (auto&& response){
+        if (!response.is2xxOK()) {
+            throw k2::dto::LogStreamBasePersistError("unable to persist metadata");
+        }
+        _create = true;
+        return seastar::make_ready_future<>();
+    });
+}
+seastar::future<std::tuple<Status, String, uint32_t> > 
+LogStreamBase::append(Payload payload, String plogId, uint32_t offset){
+    if (!_create){
+        return seastar::make_ready_future<std::tuple<Status, String, uint32_t> >(std::tuple<Status, String, uint32_t>(Statuses::S404_Not_Found("LogStreamBase does not created"), "", 0));
+    }
+    if (_current_plogId != plogId || _usedPlogInfo[_current_plogId].sealed || _usedPlogInfo[_current_plogId].currentOffset != offset){
+        return seastar::make_ready_future<std::tuple<Status, String, uint32_t> >(std::tuple<Status, String, uint32_t>(Statuses::S403_Forbidden("LogStreamBase append request information inconsistent"), "", 0));
+    }
+    return append(std::move(payload));
+};
+
+
+seastar::future<std::tuple<Status, String, uint32_t> > 
+LogStreamBase::append(Payload payload){
+    if (!_create){
+        return seastar::make_ready_future<std::tuple<Status, String, uint32_t> >(std::tuple<Status, String, uint32_t>(Statuses::S404_Not_Found("LogStreamBase does not created"), "", 0));
+    }
+    String plogId = _current_plogId;
+
+    if (_usedPlogInfo[plogId].currentOffset + payload.getSize() > PLOG_MAX_SIZE){
+        // switch to a new plog
+        return _switchPlogAndAppend(std::move(payload));
+    }
+    else{
+        uint32_t current_offset = _usedPlogInfo[plogId].currentOffset;
```

Comment:
Minor: Please add comments here, sth like
"The passed-in payload is appended into current active Plog. But currently there may be a plog switch in progress(i.e. _switched is set), and if that is case, we delay the response to this append request till switch is successfully done in the future."

Reply:
added